crates/burn-dataset/src/transform/selection.rs

```diff
@@ -0,0 +1,233 @@
+use crate::Dataset;
+use rand::SeedableRng;
+use rand::prelude::SliceRandom;
+use rand::rngs::StdRng;
+use std::marker::PhantomData;
+
+/// Generates a vector of indices from 0 to size - 1.
+///
+/// # Arguments
+///
+/// * `size` - The size of the dataset.
+///
+/// # Returns
+///
+/// A vector containing indices from 0 to size - 1.
+#[inline(always)]
+pub fn iota(size: usize) -> Vec<usize> {
+    (0..size).collect()
+}
+
+/// Generates a shuffled vector of indices up to a size.
+///
+/// # Arguments
+///
+/// * `size` - The size of the dataset to shuffle.
+///
+/// # Returns
+///
+/// A vector of shuffled indices.
+#[inline(always)]
+pub fn shuffled_indices(size: usize, rng: &mut StdRng) -> Vec<usize> {
+    let mut indices = iota(size);
+    indices.shuffle(rng);
+    indices
+}
+
+/// A dataset that selects a subset of indices from an existing dataset.
+///
+/// Indices may appear multiple times, but they must be within the bounds of the original dataset.
+pub struct SelectionDataset<D, I> {
+    dataset: D,
+    indices: Vec<usize>,
+    input: PhantomData<I>,
+}
+
+impl<D, I> SelectionDataset<D, I>
+where
+    D: Dataset<I>,
+{
+    /// Creates a new selection dataset with the given dataset and indices.
+    ///
+    /// # Arguments
+    ///
+    /// * `dataset` - The original dataset to select from.
+    /// * `indices` - A slice of indices to select from the dataset.
+    ///   These indices must be within the bounds of the dataset.
+    ///
+    /// # Panics
+    ///
+    /// Panics if any index is out of bounds for the dataset.
+    pub fn new(dataset: D, indices: Vec<usize>) -> Self {
+        let size = dataset.len();
+
+        if let Some(idx) = indices.iter().find(|&i| *i >= size) {
+            panic!("Index out of bounds for wrapped dataset size: {idx} >= {size}");
+        }
```

Comment:
I think we should not perform that check. Some datasets are huge, and doing a gazillion comparisons might not be OK. You could expose a function to perform the check or provide a checking strategy (out-of-bounds values are clipped, etc.).

Reply:
made checked and unchecked versions.